plonky3/src/circuit_builder.rs

```diff
@@ -323,3 +397,90 @@ where
         }
     }
 }
+
+impl<
+        T: FieldElementMap,
+        AB: AirBuilderWithPublicValues<F = Plonky3Field<T>> + PairBuilder + MultistageAirBuilder,
+    > MultiStageAir<AB> for PowdrCircuit<T>
+where
+    ProverData<T>: Send,
+    Commitment<T>: Send,
+{
+    fn stage_count(&self) -> usize {
+        self.constraint_system.stage_widths.len()
+    }
+
+    fn stage_trace_width(&self, stage: u32) -> usize {
+        self.constraint_system.stage_widths[stage as usize]
+    }
+
+    fn stage_challenge_count(&self, stage: u32) -> usize {
+        self.constraint_system
+            .challenges
+            .iter()
+            .filter(|c| c.stage == stage)
+            .count()
+    }
+}
+
+impl<T: FieldElementMap> NextStageTraceCallback<T::Config> for PowdrCircuit<T>
+where
+    ProverData<T>: Send,
+    Commitment<T>: Send,
+{
+    // this wraps the witgen callback to make it compatible with p3:
+    // - p3 passes its local challenge values and the stage id
+    // - it receives the trace for the next stage in the expected format, as well as the public values for this stage and the updated challenges for the previous stage
+    // - internally, the full witness is accumulated in [Self] as it's needed in order to call the witgen callback
+    fn compute_stage(
+        &self,
+        trace_stage: u32,
+        previous_stage_challenge_values: &[Plonky3Field<T>],
+    ) -> CallbackResult<Plonky3Field<T>> {
+        let witness = self.witness_so_far.lock().unwrap();
+        let mut witness = witness.borrow_mut();
+
+        let previous_stage_challenges: BTreeSet<Challenge> = self
+            .constraint_system
+            .challenges
+            .iter()
+            .filter(|c| c.stage == trace_stage - 1)
+            .cloned()
+            .collect();
+        assert_eq!(
+            previous_stage_challenges.len(),
+            previous_stage_challenge_values.len()
+        );
+        let challenge_map = previous_stage_challenges
+            .into_iter()
+            .zip(previous_stage_challenge_values)
+            .map(|(c, v)| (c.id, T::from_p3_field(*v)))
+            .collect();
+
+        let columns_before: BTreeSet<String> =
+            witness.iter().map(|(name, _)| name.clone()).collect();
+
+        // to call the witgen callback, we need to pass the witness for all stages so far
+        *witness = {
+            self.witgen_callback.as_ref().unwrap().next_stage_witness(
+                &witness,
+                challenge_map,
+                trace_stage as u8,
+            )
+        };
+
+        // generate the next trace in the format p3 expects
+        // TODO: since the current witgen callback returns the entire witness so far, we filter out the columns we already know about. Instead, return only the new witness in the witgen callback.
```

Comment:
```suggestion
        // TODO: since the current witgen callback returns the entire witness so far,
        // we filter out the columns we already know about. Instead, return only the
        // new witness in the witgen callback.
```
line breaks